x-pack/plugins/security/server/authentication/authentication_service.ts

```diff
@@ -62,12 +69,23 @@ export interface AuthenticationServiceStart {
 export class AuthenticationService {
   private license!: SecurityLicense;
   private authenticator?: Authenticator;
+  private session?: PublicMethodsOf<Session>;
 
   constructor(private readonly logger: Logger) {}
 
-  setup({ http, license }: AuthenticationServiceSetupParams) {
+  setup({ config, http, license, buildNumber }: AuthenticationServiceSetupParams) {
     this.license = license;
 
+    // If we cannot automatically authenticate users we should redirect them straight to the login
+    // page if possible, so that they can try other methods to log in. If not possible, we should
+    // render a dedicated `Unauthenticated` page from which users can explicitly trigger a new
+    // login attempt. There are two cases when we can redirect to the login page:
+    // 1. Login selector is enabled
+    // 2. Login selector is disabled, but the provider with the lowest `order` uses login form
+    const isLoginPageAvailable =
+      config.authc.selector.enabled ||
+      shouldProviderUseLoginForm(config.authc.sortedProviders[0].type);
```

Comment:
> Really nitpicky/optional nit: in the "when login page is available" test cases, it uses isLoginPageAvailable but it only appears to exercise the second conditional here.

Good catch, will test this one as well!

> This same two-part conditional is also used in Authenticator.getLoggedOutUrl, so it might be a better idea to abstract this out to a function in a separate module so it can be easily mocked and tested for.

These may look similar, but I believe they are slightly different to justify a generalization _yet_:

* When we handle `unauthenticated` case we may not know what provider user wants to use, so we care about login selector (LS) and login form (LF) only if they are displayed to the user by default when they visit Kibana (either LS is enabled, or LF is required by _the first_ provider in the chain)
* When user logs out of Kibana, we want to show LS if it's enabled and LF only if this specific provider relies on it. The only reason why we support this weird flow is to improve experience for those users that still rely on the legacy, pre-LS `providers: [saml, basic]`-like setups. E.g. to use `basic` credentials and to not trigger SAML handshake users have to navigate to `/login` route directly, and when they log out we want to redirect them back to `/login`.

I would still keep these separate as they are right now unless you feel otherwise?

Reply:
> I would still keep these separate as they are right now unless you feel otherwise?

Sounds reasonable!